plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1890,6 +1893,54 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
         }
     }
 
+    private static void checkIfCommittedAddedFilesConflictWithCurrentOperation(TupleDomain<DeltaLakeColumnHandle> enforcedSourcePartitionConstraints, IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<AddFileEntry> addedFiles = commitSummary.getIsBlindAppend().orElse(false)
+                // Do not conflict with blind appends. Blind appends can be placed before or after the current operation
+                // when backtracking which serializable sequence of operations led to the current state of the table.
+                ? List.of()
+                : commitSummary.getAddedFiles();
+
+        if (addedFiles.isEmpty()) {
+            return;
+        }
+
+        boolean readWholeTable = enforcedSourcePartitionConstraints.isAll();
+        if (readWholeTable) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files added in the modified table by concurrent write operation.");
+        }
+
+        Map<DeltaLakeColumnHandle, Domain> enforcedDomains = enforcedSourcePartitionConstraints.getDomains().orElseThrow();
+        Optional<AddFileEntry> unreadAddAction = addedFiles.stream()
+                .filter(addAction -> partitionMatchesPredicate(addAction.getCanonicalPartitionValues(), enforcedDomains))
+                .findFirst();
+        if (unreadAddAction.isPresent()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files were  added on the table with isolation level %s in the partition '%s' of the modified table  by concurrent write operation."
+                    .formatted(isolationLevel, unreadAddAction.get().getPartitionValues()));
+        }
+    }
+
+    private static void checkIfCommittedRemovedFilesConflictWithCurrentOperation(IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        if (commitSummary.getIsBlindAppend().orElse(false)) {
```

Comment:
After some more testing I found that concurrent `OPTIMIZE` is marked as blindAppend and have removedFiles in commitSummary, so it will fail. Should we mark it as not blindAppend?

Reply:
Correct. The `OPTIMIZE` needs to read from the table it optimizes :)